state-sync/aptos-data-client/src/aptosnet/mod.rs

```diff
@@ -146,46 +161,60 @@ impl AptosNetDataClient {
             })
     }
 
-    /// Fetches the next group of peers to poll. The group will contain: (i) any (new) peers that
-    /// have connected since the last time this method was called (i.e., the peers that have not
-    /// been polled yet); (ii) at most one prioritized peer (e.g., those that are upstream); and
-    /// (iii) at most one non-prioritized peer (i.e., those that are downstream).
-    fn fetch_peers_to_poll(&self) -> Result<Vec<PeerNetworkId>, Error> {
-        let mut peers_to_poll = vec![];
+    /// Fetches the next set of prioritized peers to poll. The group will contain: (i) any (new)
+    /// prioritized peers that have connected since the last time this method was called (i.e.,
+    /// the peers that have not been polled yet); and (ii) at most one prioritized peer.
+    fn fetch_prioritized_peers_to_poll(&self) -> Result<Vec<PeerNetworkId>, Error> {
+        let mut prioritized_peers = vec![];
 
         // Fetch the last polled high-priority peer
         if let Some(peer) = self.peer_states.write().oldest_polled_priority_peer() {
-            peers_to_poll.push(peer);
+            prioritized_peers.push(peer);
         }
 
-        // Fetch all new peers (i.e., those not yet polled)
+        // Fetch all new priority peers (i.e., those not yet polled)
         for peer in self.get_all_connected_peers()? {
-            if !self.peer_states.read().already_polled_peer(&peer) {
-                peers_to_poll.push(peer);
+            if !self.peer_states.read().already_polled_peer(&peer)
+                && self.peer_states.read().is_priority_peer(&peer)
+            {
+                prioritized_peers.push(peer);
             }
         }
 
-        // Handle regular peer polling
-        if peers_to_poll.is_empty() {
-            // Always try and poll at least one peer
-            if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                peers_to_poll.push(peer);
+        // Mark the peers as now polled
+        for peer in &prioritized_peers {
+            self.peer_states.write().mark_peer_as_polled(peer);
+        }
+
+        Ok(prioritized_peers)
+    }
+
+    /// Fetches the next set of regular peers to poll. The group will contain: (i) any (new)
+    /// regular peers that have connected since the last time this method was called (i.e.,
+    /// the peers that have not been polled yet); and (ii) at most one regular peer.
+    fn fetch_regular_peers_to_poll(&self) -> Result<Vec<PeerNetworkId>, Error> {
+        let mut regular_peers = vec![];
+
+        // Fetch the last polled regular peer
+        if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
+            regular_peers.push(peer);
```

Comment:
@JoshLind - Do we do some sort of deprioritization for slow peers or peers that are unstable?

Reply:
No. There is scoring of peers when actually requesting data, but nothing like that exists here. We can definitely add something like that in the future (and I think there's a larger reputation-style workstream that needs to happen), but I think it may be a little premature adding it now that we don't block.

Reply:
Do we know when peers are disconnected and remove them from the peer list for future polling?

Reply:
Yes, `get_all_connected_peers()` handles this. It ensures peers are connected and that they support the storage service protocol 😄

Reply:
@JoshLind - Maybe I am missing something here, but the `oldest_polled_regular_peer()` might return a disconnected peer as we don't seem to be updating the `regular_peer_queue` with all connected peers and removing the stale ones from the list?

Reply:
Amazing! Thank you @sitalkedia. I've updated the code to handle peer disconnect and also added tests. 

Note: The more I look at this code, the more I remember why I disagreed with the structure when it was originally ported. I think it needs to be heavily cleaned up, so I've added a comment to the code. Unfortunately, it's not high-priority yet, but if we keep needing to make "modifications" to this, I'll bite the bullet.